src/NLog/Targets/FileTarget.cs

```diff
@@ -630,6 +635,18 @@ protected override void Write(LogEventInfo logEvent)
 #endif
             byte[] bytes = this.GetBytesToWrite(logEvent);
 
+            // Clean up old archives if this is the first time a log record has been written to
+            // this log file and the archiving system is date/time based.
+            if (this.ArchiveNumbering == ArchiveNumberingMode.Date && this.ArchiveEvery != FileArchivePeriod.None)
+            {
+                if (this.previousFileName != fileName)
```

Comment:
How would this work when the `FileName` contains not only the date layout but some other dynamic layout renderer, for example some client id or logger? Then this condition could be `true` almost on every write.